spark_log_parser/loaders/s3.py

```diff
@@ -0,0 +1,103 @@
+import abc
+import boto3
+
+from pathlib import Path
+from urllib.parse import ParseResult, urlparse
+
+from botocore.response import StreamingBody
+
+from spark_log_parser.loaders import AbstractFileDataLoader, AbstractBlobDataLoader, AbstractLinesDataLoader, \
+    FileChunkStreamWrapper
+
+
+class S3StreamingBodyFileWrapper(FileChunkStreamWrapper):
+    """
+    Small wrapper around botocore.StreamingBody that exposes a lines iterator, in keeping with "pythonic" conventions.
+    botocore.StreamingBody by default iterates over raw chunks of the input stream vs. in a line-delimited manner. This
+    makes some sense, in general, since large files may not have newlines to split on. But as we expect to largely be
+    dealing with line-delimited eventlog files, this default makes more sense for us to use.
+    """
+
+    def __init__(self, body: StreamingBody):
+        # Since this stream is intended to be handed off to other classes for manipulation (i.e. unzipping), we set
+        # keepends=True because various decoding libraries may throw decode errors if we remove parts of the binary
+        # data stream
+        #
+        # Note - We need to watch out for this, though, as this means that iterating lines will give us binary buffers
+        #  with `\n` at the end still. So if we are fetching an uncompressed file from S3, whatever is parsing those
+        #  lines may need to be tolerant of that, OR we need to configure this flag properly. This would be pretty
+        #  do-able by inspecting file extensions in AbstractS3FileDataLoader.load_item, but is not currently done.
+        #  Right now, we may hit this case when loading raw .json files from S3, but json.loads(line) is tolerant of
+        #  trailing newline characters. test_parse_s3.test_raw_log_from_s3 exercises this code path for raw .json files
+        self._chunks = body.iter_lines(chunk_size=1024 * 1024, keepends=True)
```

Comment:
We do need to keep them, unfortunately... not for our logic, but for the libraries we're calling into to unzip/decompress stuff. Unzipping a zip archive with parts of the binary stream removed will, for example, raise an exception that it's an invalid zip file. If I switch this to `False` and re-run the tests, I get these failures in the S3 tests -

```python
FAILED tests/test_parse_s3.py::test_emr_log_from_s3[event_log_file_archive0-airlinedelay/jb-42K1E16/-s3://sync-test-artifacts/airlinedelay/jb-42K1E16/emr.zip] - orjson.JSONDecodeError: unexpected character: line 1 column 135 (char 134)
FAILED tests/test_parse_s3.py::test_emr_log_from_s3[event_log_file_archive0-airlinedelay/jb-42K1E16/-s3://sync-test-artifacts/airlinedelay/jb-42K1E16/] - orjson.JSONDecodeError: unexpected character: line 1 column 135 (char 134)
FAILED tests/test_parse_s3.py::test_parsed_log_from_s3[event_log_file_archive0-/foo/bar-baz/-s3://sync-test-artifacts/foo/bar-baz/similarity_parsed.json.gz] - zlib.error: Error -3 while decompressing data: invalid bit length repeat
FAILED tests/test_parse_s3.py::test_parsed_log_from_s3[event_log_file_archive0-/foo/bar-baz/-s3://sync-test-artifacts/foo/bar-baz/] - zlib.error: Error -3 while decompressing data: invalid bit length repeat
FAILED tests/test_parse_s3.py::test_databricks_log_from_s3_dir[event_log_file_archive0-airlinedelay/jb-42K1E16/-s3://sync-test-artifacts/airlinedelay/jb-42K1E16/] - zlib.error: Error -3 while decompressing data: invalid block type
FAILED tests/test_parse_s3.py::test_databricks_log_from_s3_dir[event_log_file_archive0-airlinedelay/jb-42K1E16/-s3://sync-test-artifacts/airlinedelay] - zlib.error: Error -3 while decompressing data: invalid block type

```

Reply:
Is that because the line loader delegates to this? Maybe this should use `iter_chunks` instead.

Reply:
Kind of - in the line loader, we want `__iter__` to behave in a manner consistent with how other file-like objects work (i.e. `for line in f` yields actual lines of the file, not haphazard chunks, which is what `StreamingBody` does by default already). And of course we don't want to use `readlines()`, because that tends to buffer the whole file in memory. Some code paths will be tolerant of `\n` missing from that stream, for example if we just have a regular `.json` file with JSON lines. But others, like unpacking a `.zip` archive first, will not be. I can play around with this a bit though and see if it's possible, though!

Reply:
OK, I have what I think is a viable alternative. Instead of relying on `for line in file` to return the well-formed lines for us, it's possible to ensure that we always pass a `FileChunkStreamWrapper` into `read_file_stream`. This would allow us to `yield from file.iter_lines()`, like - 

```python
class AbstractLinesDataLoader(AbstractFileDataLoader, abc.ABC):
    """
    Abstract class that implements the various read_* methods from AbstractFileDataLoader in such a way
    as to yield individual lines of the underlying file source. Useful for loading e.g. JSON Lines or CSV files
    """

    cache = False

    def read_file_stream(self, file: FileChunkStreamWrapper):
        yield from file.iter_lines()
```

This would allow us to just use the default chunking behavior from `StreamingBody`.

Reply:
This feels a bit better anyway, because while `for line in f` may be the "pythonic convention", it's not explicit, and as we're seeing, not always conformed to!

Reply:
...and by passing in our own wrapper, we can more easily control that "conformance"

Reply:
I was able to get this to work with `iter_chunks` by making the changes that I outlined in the comments above ^^ :+1: